packages/devkit/src/utils/ignore-path.ts

```diff
@@ -0,0 +1,42 @@
+/**
+ * This function can be used to ignore paths when running a task (generator, etc).
+ * It checks if a provided path is included in the ignorePaths list.
+ * It uses the format of the inputs, all paths must start
+ * with either {workspaceRoot} or {projectRoot}.
+ *
+ * @param path the path of the file to check against the ignore list
+ * @param ignorePaths the list of paths to ignore - all paths
+ *                    must start with either {workspaceRoot} or {projectRoot}
+ * @param projectRoot the root of the current project
+ * @returns boolean indicating if the path should be ignored
+ */
+
+import minimatch = require('minimatch');
+
+export function ignorePath(
+  path: string,
+  ignorePaths: string[],
+  projectRoot: string
+): boolean {
+  return ignorePaths.some((pattern) => {
+    if (
+      !pattern.startsWith('{projectRoot}/') &&
+      !pattern.startsWith('{workspaceRoot}/')
+    ) {
+      throw new Error(
+        [
+          `"${pattern}" in your ignorePaths is an invalid path.`,
+          'All paths have to start with either {workspaceRoot} or {projectRoot}.',
+          'For instance: "{projectRoot}/**/not-stories/**" or "{workspaceRoot}/**/**/not-stories/**".',
+        ].join('\n')
+      );
+    } else {
+      return minimatch(
+        path,
+        pattern.startsWith('{projectRoot}/')
+          ? pattern.replace('{projectRoot}', projectRoot)
+          : pattern.substring(16)
+      );
+    }
+  });
+}
```

Comment:
Should we just use the `ignore` package for this? I understand the portion regarding normalizing `{projectRoot} / {workspaceRoot}`, but I almost feel like its a separate concern from handling the ignore element.

If we don't update (or remove) this method in favor of `ignore`, I feel like we should rename it. `ignorePath` sounds like a utility that may read .gitignore / .nxignore to determine if the path was ignored... This method could be named something like `pathMatch` since it doesn't directly ignore the path (that's left to the consumer of the function).

Reply:
I thought of the `ignore` package, but I could not really see how I could use it in this case, to be honest.... Do you have any suggestions as to how it could apply it in this case?

I see the concern about the naming, you're 100% correct!! Thanks!!

Reply:
There's a similar function in [hasher](https://github.com/nrwl/nx/blob/master/packages/nx/src/hasher/hasher.ts#L334) but I could not see how I could re-use it either, since it's sort of taking into account more things, and also wants to return a different result...

Reply:
Something like this is what I was thinking:

```typescript
export function ignorePath(
  path: string,
  ignorePaths: string[],
  projectRoot: string
): boolean {
  const ig = ignore()
    .add(
      pathsToIgnore.map((path) =>
        path.startsWith('{projectRoot}/')
          ? path.replace('{projectRoot}', root)
          : path.substring(16)
      )
    )
  return ig.ignores(path)
}
```

Reply:
(disregarding naming etc)

Reply:
If we were to do something like that, I'd split the `{projectRoot}` handling out though. 

```typescript
// nx/src/utils/path.ts

/**
  * Replaces workspaceRoot and projectRoot tokens in a string.
  */
export function resolveWorkspacePath(path: string, projectRoot: string) {
  return path
                 .replace('{projectRoot}', projectRoot)
                 .replace('{workspaceRoot}', workspaceRoot)
}
```

I wouldn't export the usage of `ignore` in @nrwl/devkit, since its just a small line item after the above is pulled out. In each place that you were calling ignorePath, you could do:

```typescript
const ignoredPaths = ignore().add(pathsToIgnore.map((path) => resolveWorkspacePath(path, projectRoot));

visitNotIgnoredFiles(tree, (filePath) => {
  if (!ignoredPaths.ignores(filePath) {
    // ...
  }
})

```

Reply:
Ah I see what you mean. Instead of minimatch to just use `ignores` but sort of keep the same logic. It just seems to me however that in this case we're doing an extra `map`, no?

Reply:
Oh, also I realized that `minimatch` uses a slightly different matching pattern than `ignore`, which uses the gitignore spec.